### PR TITLE
Move factory to a new file in common

### PR DESCRIPTION
**Pull request by jroweboy**
As we add more devices and different types of factories, its useful to
not repeat the same plumbing code for factories, so this refactors the
input factory to common.

This commit did not change the factory for cameras as that is handled
slightly differently and would be fine to put in a follow up after the
still camera is merged. See #3451 for the still camera.

This is the first PR in relation to adding a frontend HLE applet interface. I plan to extend this code to add support for frontend driven applets, and wanted to reuse the existing factory code. Please review quickly so I can push out the other code as well.

The goal is to allow frontends to define their own applet callbacks, for instance, the QT frontend can have a software keyboard that allows users to type text with their keyboard, or any other frontend can draw a keyboard to the screen and use controller input to choose the keys.

---
This change is [Reviewable](https://reviewable.io/reviews/citra-emu/citra/3553)

### CODE REVIEW

File: src/input_common/sdl/sdl.cpp [hunk 1]
```diff
@@ -238,17 +239,17 @@ void Init() {
         LOG_CRITICAL(Input, "SDL_Init(SDL_INIT_JOYSTICK) failed with: %s", SDL_GetError());
     } else {
         using namespace Input;
```
Comment (B3n30) [inline src/input_common/sdl/sdl.cpp]: can be removed

File: src/input_common/sdl/sdl.cpp [hunk 2]
```diff
@@ -238,17 +239,17 @@ void Init() {
...
-        RegisterFactory<AnalogDevice>("sdl", std::make_shared<SDLAnalogFactory>());
+        Common::RegisterFactory<ButtonDevice>("sdl", std::make_shared<SDLButtonFactory>());
+        Common::RegisterFactory<AnalogDevice>("sdl", std::make_shared<SDLAnalogFactory>());
         initialized = true;
     }
 }
 
 void Shutdown() {
     if (initialized) {
         using namespace Input;
```
Comment (B3n30) [inline src/input_common/sdl/sdl.cpp]: dito

**Comment by wwylele** [comment]
From Discord:

> wwylele / 白疾風 - Today at 7:54 PM
> I have a concern that this might lead to abuse of the fixed factory template...
> like, I don't really expect applet implementation need the ParamPackage thing, as it doesn't need to be configurable other than the main factory choice
> I mean, each actual factory group has their own features. They might be configured in a different way (probably by multiple member functions), might contain more state (by member variables)
> This code make it like any factory pattern should make use of it, whereas in reality they vary from each other in many ways
> jroweboy - Today at 8:00 PM
> what would you recommend instead?
> wwylele / 白疾風 - Today at 8:01 PM
> just... make your own "applet factory" without sharing the code with input devices? Do you really need ParamPackage?
> The input core needs it because the input system needs to be extendable, as in one can easily make a new implementation for a new controller
> I don't expect that applet system needs that extendability and people would write more and more implementation for it
> jroweboy - Today at 8:05 PM
> making my own factory was a copy paste of the factory code, just without the full param package. its the same thing that camera does, and its kinda pointless to keep copy pasting the factory code for every new kind of frontend device <-> core communication
> wwylele / 白疾風 - Today at 8:06 PM
> Ah probably I should go back to this question first:
> do you really need a factory?
> wwylele / 白疾風 - Today at 8:08 PM
> the code I imagined is this simple:
> class AppletInterface {...};
> class LLEApplet: AppletInterface{...};
> std::xxx_ptr<AppletInterface> lle = std::make_xxx<LLEApplet>();
> std::xxx_ptr<AppletInterface> hle = /*set by frontend*/
> ...
> if (settings.use_lle){
>     lle->foo();
> } else {
>     hle->foo();
> }

**Comment by jroweboy** [comment]
Closing as I have a better understand of what is the expected way to do this

jroweboy - Today at 12:23 PM
so how would the HLE applet code in know that it should launch any keyboard from the frontend instead of the default keyboard that is in core? (In this case the default keyboard simply types "Citra" and pushes the second button)
and i answered that by saying "Read the value in settings for hle_keyboard and create the keyboard device from the factory"
wwylele / 白疾風 - Today at 12:24 PM
core just call frontend "Please bring up a proper keyboard" via a solid, non-virtualized interface (or probably a callback)
and let frontend choose the keyboard itself
jroweboy - Today at 12:25 PM
but then the core doesn't have a way to call the default keyboard
wwylele / 白疾風 - Today at 12:25 PM
?
Sorry, I need to make something clear first,
when to call the default keyboard again?
jroweboy - Today at 12:27 PM
the default keyboard is used if there is no keyboard for the frontend. are we fine adding a new requirement, that frontends have to have a hle applet? i don't like that.
wwylele / 白疾風 - Today at 12:27 PM
Then the frontend returns an error via the interface mentioned above "Sorry I don't have a proper keyboard" and core will know it
jroweboy - Today at 12:27 PM
i'm fine with that i suppose
B3N30 - Today at 12:27 PM
if no callback is registered use the default keyboard?
wwylele / 白疾風 - Today at 12:28 PM
^ that's another answer
basically, frontend only need to register ONE interface to the core
the rest is the implementation detail in frontend
